vj4/handler/contest.py

```diff
@@ -402,40 +412,22 @@ class ContestEditHandler(base.Handler, ContestStatusMixin):
     tdoc = await contest.get(self.domain_id, tid)
     if not self.own(tdoc, builtin.PERM_EDIT_CONTEST_SELF):
       self.check_perm(builtin.PERM_EDIT_CONTEST)
-    if self.is_live(tdoc) or self.is_done(tdoc):
-      if begin_at_date:
-        raise error.ValidationError('begin_at_date')
-      if begin_at_time:
-        raise error.ValidationError('begin_at_time')
-      begin_at = tdoc['begin_at']
-    else:
-      try:
-        begin_at = datetime.datetime.strptime(begin_at_date + ' ' + begin_at_time, '%Y-%m-%d %H:%M')
-        begin_at = self.timezone.localize(begin_at).astimezone(pytz.utc).replace(tzinfo=None)
-      except ValueError as e:
-        raise error.ValidationError('begin_at_date', 'begin_at_time')
+    try:
+      begin_at = datetime.datetime.strptime(begin_at_date + ' ' + begin_at_time, '%Y-%m-%d %H:%M')
+      begin_at = self.timezone.localize(begin_at).astimezone(pytz.utc).replace(tzinfo=None)
+    except ValueError as e:
+      raise error.ValidationError('begin_at_date', 'begin_at_time')
     end_at = begin_at + datetime.timedelta(hours=duration)
     if begin_at >= end_at:
       raise error.ValidationError('duration')
-    # not allow removing existing problems
-    pid_set = set(map(document.convert_doc_id, pids.split(',')))
-    if self.is_live(tdoc) or self.is_done(tdoc):
-      old_set = set(tdoc['pids'])
-      if not old_set <= pid_set:
-        raise error.ValidationError('pids')
-    pids = list(pid_set)
+    pids = list(set(map(document.convert_doc_id, pids.split(','))))
     pdocs = await problem.get_multi(domain_id=self.domain_id, doc_id={'$in': pids},
                                     fields={'doc_id': 1}) \
                          .sort('doc_id', 1) \
                          .to_list()
-    exist_pids = [pdoc['doc_id'] for pdoc in pdocs]
-    if len(pids) != len(exist_pids):
-      for pid in pids:
-        if pid not in exist_pids:
-          raise error.ProblemNotFoundError(self.domain_id, pid)
-    # TODO(twd2): further checks for editing?
     await contest.edit(self.domain_id, tdoc['doc_id'], title=title, content=content,
                        rule=rule, begin_at=begin_at, end_at=end_at, pids=pids)
     for pid in pids:
       await problem.set_hidden(self.domain_id, pid, True)
+    await contest.recalc_status(self.domain_id, tdoc['doc_id'])
```

Comment:
To improve performance, only editing `rule`, `begin_at`, `pids` should trigger this recalc.